tests/test_generate_report.py

```diff
@@ -0,0 +1,183 @@
+        if os.path.exists(name) and os.path.isfile(name):
+            os.remove(name)
+
+    def test_not_implemented(self):
+        """test to make sure analysis module fails expectedly when functions are called directly"""
+        a = AnalysisModule()
+        with pytest.raises(NotImplementedError):
+            a.process_dict()
+
+        with pytest.raises(NotImplementedError):
+            str(a)
+
+    def test_svm(self):
+        """test the process_json function when the target model is an SVM"""
+        json_formatted = self.get_test_report()
+
+        f = FinalRecommendationModule(json_formatted)
+        returned = f.process_dict()
+
+        self.assertEqual(len(returned["score_descriptions"]), 0)
+
+        json_formatted["model"] = "SVC"
+        f = FinalRecommendationModule(json_formatted)
+        returned = f.process_dict()
+
+        self.assertIn("Model is SVM", returned["score_descriptions"][0])
+
+    def test_min_samples_leaf(self):
+        """test the process_json function when the target model is a random forest"""
+        json_formatted = self.get_test_report()
+
+        f = FinalRecommendationModule(json_formatted)
+        returned = f.process_dict()
+
+        self.assertEqual(len(returned["score_descriptions"]), 0)
+
+        json_formatted["model_params"]["min_samples_leaf"] = 2
+        f = FinalRecommendationModule(json_formatted)
+        returned = f.process_dict()
+
+        self.assertIn("Min samples per leaf", returned["score_descriptions"][0])
+
+    def test_statistically_significant(self):
+        """test the statistically significant AUC p-values check in FinalRecommendationModule"""
+        json_formatted = self.get_test_report()
+        json_formatted["attack_experiment_logger"]["attack_instance_logger"] = {}
+
+        metrics_dict = {
+            "P_HIGHER_AUC": 0.001,
+            "AUC": 0.8,
+            "ACC": 0.7,
+            "FDIF01": 0.2,
+            "PDIF01": 1.0,
+            "FPR": 1.0,
+            "TPR": 0.1,
+        }
+
+        for i in range(10):
+            json_formatted["attack_experiment_logger"]["attack_instance_logger"][
+                "instance_" + str(i)
+            ] = metrics_dict
+
+        f = FinalRecommendationModule(json_formatted)
+        returned = f.process_dict()
+
+        self.assertIn(
+            ">10% AUC are statistically significant", returned["score_descriptions"][0]
+        )
+        self.assertIn("Attack AUC > threshold", returned["score_descriptions"][1])
+
+    def test_univariate_metrics_module(self):
+        """test the SummariseUnivariateMetricsModule"""
+        json_formatted = self.get_test_report()
+        f = SummariseUnivariateMetricsModule(json_formatted)
+        _ = f.process_dict()
+
+    def test_auc_pvals_module(self):
+        """test the SummariseAUCPvalsModule"""
+        json_formatted = self.get_test_report()
+        f = SummariseAUCPvalsModule(json_formatted)
+        _ = f.process_dict()
```

Comment:
Happy with this for now. After this release I think we should actually test the output of these. If we have a static json output produced by #126 within the repo, we can use it to test that the modules are also giving the correct answers (and it will also pick up any module output formatting changes that creep in).